feature: expose shoko events through SignalR + more

revam opened the pull request:
expose the Shoko events through SignalR and refactor the plugin file events.

revam commented:
The file events are tested on my machine:tm:
![image](https://user-images.githubusercontent.com/7761729/184503243-d5f0488d-0141-4424-8f3a-b6f4962c2330.png)
![image](https://user-images.githubusercontent.com/7761729/184503253-5baa97cc-4476-4e0f-af59-eafcafbe9f6a.png)

Cazzar commented:
Just a noticed pattern - it seems very often you seem to be adding extra changes unrelated to what you are doing.
For example
1. adding the docblocks to random objects, while it's not horrible of an idea - it makes it more complex to review.
2. Why are you adding the `: EventArgs` to some of the already existing event args.
If these are done for good reason, it's usually worth noting into the PR description

revam commented:
1. I didn't add any doc-blocks to any "random" objects in this PR — every added doc-block is followed by an added or updated field. I will reverse back the order of the fields in `FileEventArgs` though, so it's easier to see what actually changed.
2. Because to me is an `*EventArgs` not extending `System.EventArgs` (or a derived class) the same as an `*Exception` not extending `System.Exception`. Both patterns work in practice, but personally I don't like it.

Will try to write more details in the description next time.

revam commented:
Fixed up the order of the existing fields, removed some empty lines at the beginning of the new files, and removed unused dependencies/imports in the new files.

Cazzar commented:
> 1. I didn't add any doc-blocks to any "random" objects in this PR — every added doc-block is followed by an added or updated field. I will reverse back the order of the fields in `FileEventArgs` though, so it's easier to see what actually changed.
> 2. Because to me is an `*EventArgs` not extending `System.EventArgs` (or a derived class) the same as an `*Exception` not extending `System.Exception`. Both patterns work in practice, but personally I don't like it.
> 
> Will try to write more details in the description next time.

For future reference, I would suggest discussing changes you want to make "because you don't like it" because there may be unknown side-effects, or reasons it has been done explicitly.

The reason I said random, is when looking at the issue itself, and documentation you had included, a lot of it seemed unrelated, to the scope of "Expose x events"

Cazzar commented on Shoko.Server/API/SignalR/Models/EpisodeInfoUpdatedEventSignalRModel.cs:
```diff
@@ -0,0 +1,32 @@
+            EpisodeID = ((SVR_AnimeEpisode)eventArgs.EpisodeInfo).AnimeEpisodeID;
+            SeriesID = series.AnimeSeriesID;
+            GroupID = series.AnimeGroupID;
+        }
```
This as a whole seems like a N+1 issue, why do we need to have this lookup - is it better to have it in the EventArgs model itself?

revam replied:
The plugin abstraction uses the SVR_AniDB_Anime for it's IAnime implementation, The SVR_AniDB_Anime model does not contain the shoko ids. In my opinion if we want to fix having to look up the series from the anime id then we need to better align the plugin abstractions with how shoko internally models the data.

Cazzar replied:
That would ultimately solve the N+1 issue that has been created here, because what is now happening is, we look up the details, create the interface details, pass it down to the event and look it back up.

Ultimately this becomes a potential problem for performance issues, and we should be grabbing it only once.

If you need more detail on the concept of N+1 issues: https://medium.com/doctolib/understanding-and-fixing-n-1-query-30623109fe89 while in rails, the concepts are entirely the same, and will become more evident when we move to EF, and don't really have the caches in the long run.

Cazzar commented on Shoko.Server/API/SignalR/Models/FileHashedEventSignalRModel.cs:
```diff
@@ -0,0 +1,11 @@
+        {
+        }
+    }
+}
```
Unless this is adding some sort of value, it is probably worth just passing in the base class.

revam replied:
We can add fields to it in the future if needed. Yes we can change it.

Cazzar replied:
Yeah, I would just add it when it's needed, not trying to plan too far in the future

Cazzar commented on Shoko.Server/API/SignalR/ShokoEventEmitter.cs:
```diff
@@ -0,0 +1,60 @@
+        private async void OnEpisodeUpdated(object sender, EpisodeInfoUpdatedEventArgs e)
+        {
+            await Hub.Clients.All.SendAsync("EpisodeUpdated", new EpisodeInfoUpdatedEventSignalRModel(e));
+        }
```
For this, why aren't we just sending in the EventArgs directly for most of these, it seems like pointless encapsulation?

revam replied:
Because in both Web UI and Shokofin we don't need the extra data exposed to the plugins sent to the clients. And if we do need the data in the future then we can fetch it using the IDs sent instead.

Cazzar replied:
Having extra data isn't a bad thing, you can just use what is needed

Cazzar reviewed (commented):
Further to this, when making any changes to Shoko.Plugin.Abstractions, it will force any person who uses daily and a custom plugin to rebuild it, because the dependency changes, so I would suggest only making necessary changes, or batching them together.

revam commented:
> Further to this, when making any changes to Shoko.Plugin.Abstractions, it will force any person who uses daily and a custom plugin to rebuild it, because the dependency changes, so I would suggest only making necessary changes, or batching them together.

1. I just batched them togther.
2. I'm using my plugin built for 4.1.1 and it's working fine — even on the daily built earlier today — because I didn't touch any of the renamer events, only the new (and hopefully still unused) events introduced in (I think it was) 4.1.2.

Cazzar commented:
> > Further to this, when making any changes to Shoko.Plugin.Abstractions, it will force any person who uses daily and a custom plugin to rebuild it, because the dependency changes, so I would suggest only making necessary changes, or batching them together.
> 
> 1. I just batched them togther.
> 2. I'm using my plugin built for 4.1.1 and it's working fine — even on the daily built earlier today — because I didn't touch any of the renamer events, only the new (and hopefully still unused) events introduced in (I think it was) 4.1.2.

2. I have seen this happen too many times before, I was just mentioning in terms of previous experience
